File: src/main/scala/io/koju/autopos/party/domain/Customer.scala
package io.koju.autopos.party.domain

import java.lang.Long
import javax.persistence.GenerationType.SEQUENCE
import javax.persistence._
import javax.validation.constraints.{NotNull, Size}

import io.koju.autopos.kernel.domain.AuditableBaseEntity

import scala.beans.BeanProperty

@Entity
@Table(name = "customer")
class Customer extends AuditableBaseEntity {

  @Id
  @SequenceGenerator(name = "customer_id_seq", sequenceName = "customer_id_seq", allocationSize = 1)
  @GeneratedValue(strategy = SEQUENCE, generator = "customer_id_seq")
  @BeanProperty
  var id: Long = _

  @NotNull
  @Size(min = 2, max = 100)
  @Column(name = "name", length = 100, nullable = false, unique = true)
  @BeanProperty
  var name: String = _

  @Size(max = 250)
  @Column(name = "remarks", length = 250)
  @BeanProperty var remarks: String = _
}

File: src/main/kotlin/kotlin/com/kapilkoju/autopos/party/domain/Customer.kt
package com.kapilkoju.autopos.party.domain

import com.kapilkoju.autopos.kernel.domain.AuditableEntity
import javax.persistence.*
import javax.validation.constraints.NotNull
import javax.validation.constraints.Size


@Entity
@Table(name = "customer")
@NamedEntityGraph(name = Customer.Graph.detail,
        attributeNodes = [NamedAttributeNode("phoneNumbers")])
data class Customer(

        @NotNull
        @Size(min = 2, max = 100)
        @Column(name = "name", length = 100, nullable = false, unique = true)
        val name: String,

        @ElementCollection
        @CollectionTable(name = "customer_phone_numbers")
        @Column(name = "phone_number")
        val phoneNumbers: List<String>? = ArrayList(),

        @Size(max = 250)
        @Column(name = "remarks", length = 250)
        val remarks: String?

) : AuditableEntity() {

    companion object Graph {
        const val detail = "Customer.detail"
    }

}


